builder/builder.go

```diff
@@ -0,0 +1,390 @@
+package builder
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Builder struct {
+	spacing string // optional spacing [s]
+	newline string
+	err     error
+
+	locals            []local
+	inputs, selectors []Expression
+	variants          []variant
+	patterns          []any
+}
+
+func New() *Builder {
+	return &Builder{
+		newline: "\n",
+		spacing: " ",
+	}
+}
+
+func (b *Builder) Build() (string, error) {
```

Comment:
The target consumers and intent are different. We don't want consumers to use AST to build MF2 messages.

* Parse() parses MF2 into AST (we don't have the intent to do AST.String(), we did it by inertia)
* Builder makes it easy and safe to construct MF2 messages

E.g. we don't want to burden consumers on how to handle quoted/unquoted literals. The consumer does NOT care.

Builder helps us to convert from different i18n formats into MF2.

Reply:
its not about that.

like AST, has Stringer implementation, right ? AST -> mf2 string.

my question is:
why cant builder have additional field unexported ast, to call that AST Stringer implementation, instead of defining new Stringer implementation for those new structs, which are doing the same thing

Reply:
my main concern is the following:
we have almost perfect nodes defined in AST, with each node, representing itself as string. But here instead of using those nodes, we are creating derivatives of them, almost the same but a little different. 
My main concerns:
a) Instead of maintaining only ast.go nodes, we must also maintain builder.go nodes, in case the messageformat abnf changes
b) Instead of using already implemented node->string, we are reimplementing the same thing for derivativeNode -> string, which strengthens my point a)